Added new ecs_task_facts module

kaczynskid opened the pull request:
##### Issue Type:
- New Module Pull Request
##### Plugin Name:

cloud/amazon/ecs_task_facts
##### Ansible Version:

```
ansible 2.1.0 (devel 2db3f290ba) last updated 2016/02/24 10:06:09 (GMT +200)
  lib/ansible/modules/core: (detached HEAD 7162623e86) last updated 2016/02/24 10:09:54 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD ecs_servic) last updated 2016/02/24 12:04:25 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
```
##### Summary:

This module complements Amazon ECS support in Ansible by providing possibility to list ECS tasks either by their IDs or ARNs or by search criteria including task definition family, name of the service the task was created within and expected task status.

robynbergeron commented:
Thanks @kaczynskid for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

kaczynskid commented:
@Java1Guy would you mind reviewing this one as well?

Java1Guy commented:
@kaczynskid it is interesting you created this -- I had also!  But for whatever reason, the reviewers rejected it when I submitted my PR - i think they just didn't have time to review.. i don't really know.
Anyway, you can see my version in my repo in the ecs_task_module branch of my extras -- https://github.com/Java1Guy/ansible-modules-extras/blob/ecs_task_module/cloud/amazon/ecs_task_facts.py

Have a quick look at that...  A quick look says yours is very similar.  Probably the biggest difference is that I expect the parameters in "Ansible" style (with '_'s) and created a function to change them to the boto parameters (transmogrify).

See what you think -- make any changes you feel appropriate and I'll have another look in a few days.

gregdek commented:
Thanks @kaczynskid for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@kaczynskid A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@kaczynskid Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@kaczynskid A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

kaczynskid commented:
Thank you @Java1Guy. I looked at your code and added some more parameters that you had there.

I also rebased this branch against `stable2.1` branch, bu I cannot edit the pull request. Can this request be edited by someone to merge into `stable2.1`? Or should I close this one and create a separate pull request? Or chould I rebase this against current `devel`?

kaczynskid commented:
OK, so I found out that pull request target branch cannot be changed. Therefore I rebased the changes against current develop and squished the commits. This is still compatible with stable-2.1 and I'd love to see it merged there as well and I'll create a separate pull request for that when this is merged.

kaczynskid commented:
@Java1Guy @daniel-rhoades @jianyuan @pvwichen - ready_for_review

gregdek commented:
Thanks @kaczynskid for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

kaczynskid commented:
@gregdek I looked at #2357 and would like to ask if you please can add @rmorlok as maintainer to review this module as well?

gregdek commented:
@kaczynskid he has been added as maintainer of that module as well.

rmorlok commented:
@kaczynskid I'll take a look at this as well, but this weekend is the earliest I'd be able to play around with it

kaczynskid commented:
@rmorlok friendly ping :)

rmorlok commented on cloud/amazon/ecs_task_facts.py:
```diff
@@ -0,0 +1,287 @@
+        fn_args = dict()
+
+        # apply non-exclusive parameters
+        if cluster and cluster is not None:
```
Because `None` is falsey, this duplicate check is redundant. Same for subsequent if statements below.

rmorlok commented on cloud/amazon/ecs_task_facts.py:
```diff
@@ -0,0 +1,287 @@
+
+    argument_spec = ec2_argument_spec()
+    argument_spec.update(dict(
+        details=dict(required=False, choices=['true', 'false']),
```
This seems like it should of `type="bool"` instead of choices.

rmorlok commented:
needs_revision One error I noticed is that if you have details=true and the query does not return any tasks an exception get's thrown rather than giving a graceful failure or empty set. It works fine if the details aren't requested.

kaczynskid commented:
@rmorlok thx for tour comments :) did you had a moment to review my fixes after your last comment?

kaczynskid commented:
@rmorlok any chance you could have a look - I applied fixes after your comments - thx :)

rmorlok commented:
Hi @kaczynskid

I notice that you changed `get_tasks_list(...)` to return tasks (which might be None) when it catches an error. Is that the correct behavior?

If I do something like:

```
---
- hosts: local
  tasks:
   - ecs_task_facts:
      cluster: this-cluster-exists
      service: bad-service
      details: false
```

The check for the service fails, but instead of filtering the tasks to an empty set, it leaves the tasks in place that it found for the cluster. It seems like `get_tasks_list(...)` should be returning an empty set when it gets a failure. Do you agree? Was there a reason you changed it?

rmorlok commented:
Also, the changes also seem to make this fail for if the only filter isn't valid.

This would throw an exception regardless of the value for `details`:

```
---
- hosts: local
  tasks:
   - ecs_task_facts:
      cluster: does-not-exist-cluster
      details: false
```
